### PR TITLE
better interface with alertmanager

### LINKED ISSUE #47: config_hash calculated incorrectly
https://github.com/canonical/prometheus-operator/blob/56c820f5cd093846fc0fe0f642e087f5782d30d9/src/charm.py#L91

`config_hash` may be the same value even for different strings:
```
>>> prometheus_config1 = "{'global': {'scrape_interval': '1m', 'scrape_timeout': '10s', 'evaluation_interval': '1m'}, 'scrape_configs': [{'job_name': 'prometheus', 'scrape_interval': '5s', 'scrape_timeout': '5s', 'metrics_path': '/metrics', 'honor_timestamps': True, 'scheme': 'http', 'static_configs': [{'targets': ['localhost:9090']}]}], 'alerting': {'alertmanagers': [{'static_configs': [{'targets': ['10.1.157.127', '10.1.157.71', '10.1.157.74']}]}]}}"
>>> prometheus_config2 = "{'global': {'scrape_interval': '1m', 'scrape_timeout': '10s', 'evaluation_interval': '1m'}, 'scrape_configs': [{'job_name': 'prometheus', 'scrape_interval': '5s', 'scrape_timeout': '5s', 'metrics_path': '/metrics', 'honor_timestamps': True, 'scheme': 'http', 'static_configs': [{'targets': ['localhost:9090']}]}], 'alerting': {'alertmanagers': [{'static_configs': [{'targets': ['10.1.157.127', '10.1.157.71', '10.1.157.72', '10.1.157.74']}]}]}}"
>>> str(hashlib.md5(str(prometheus_config1).encode('utf-8')))
'<md5 HASH object @ 0x7f54ddcdfa10>'
>>> str(hashlib.md5(str(prometheus_config2).encode('utf-8')))
'<md5 HASH object @ 0x7f54ddcdfa10>'
```
A call to `.hexdigest()` is missing:
```
>>> hashlib.md5(str(prometheus_config1).encode('utf-8')).hexdigest()
'48e7ae9dce66f3022a3b5342e002c356'
>>> hashlib.md5(str(prometheus_config2).encode('utf-8')).hexdigest()
'6e1dcd35bee04b36d76674128782230b'
```

### PR DESCRIPTION
**Pull request by sed-i**
Updates to prometheus, to go with corresponding [changes in alertmanager](https://github.com/canonical/alertmanager-operator/pull/7).

### CODE REVIEW

File: src/charm.py
Comment (balbirthomas) [inline src/charm.py:98]: Thanks. Although it may be more appropriate if a  separate PR was created just for this issue so that it can be merged without waiting for this draft PR and besides the two change sets are unrelated.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [inline lib/charms/alertmanager_k8s/v0/consumer.py:56]: I am curious to understand why you choose to make the consumer (rather than the provider) responsible for aggregating "public address" of all Alertmanager peers. In the same vein I would be interested in your rationale for using stored state in the charm library to store this list rather than application data.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [inline lib/charms/alertmanager_k8s/v0/consumer.py:68]: Doesn't this mean there will be two `available` events emitted for each joined unit (since the ConsumerBase also emits one) ? Is that the intention ?  Instead would it be possible to just ensure that `get_cluster_info()` provides the required list of addresses to Prometheus when `Prometheus.AlertmanagerConsumer` raises `available` ?

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [inline lib/charms/alertmanager_k8s/v0/consumer.py:52]: What should happen when alertmanager is upgraded and no longer compatible with Prometheus. Can this happen ?

File: src/charm.py
Comment (balbirthomas) [inline src/charm.py:40]: Why not just delete it instead of commenting it out ? Or are you planning to uncomment it in a future revision of this draft ?

File: src/charm.py
Comment (balbirthomas) [inline src/charm.py:56]: If I remember correctly you can also pass None for version if there is no restriction on version. Please do check this if you want to do it that way. Also I presume you have indeed ascertained that any Prometheus version will work with any version of Alertmanager.

File: src/charm.py
Comment (balbirthomas) [inline src/charm.py:57]: Very minor style comment. Do you think of the composition here as "a Prometheus object contains annAlertmanager Library" object or "a Prometheus object contains and Altermanager Consumer" object or even just and "Alertmanager" object.

File: src/charm.py
Comment (balbirthomas) [inline src/charm.py:146]: Is there any reason to store the list of alert managers both in the AlertmanagerConsumer library and the Prometheus Charm ? Why not just integrate it into `_alerting_config()` ?

File: tests/test_charm.py
Comment (balbirthomas) [inline tests/test_charm.py:68]: Same question about when not just delete rather than commenting it out. PS: I have not check test coverage since this is a draft. I trust your workmanship here.

File: tests/test_charm.py
Comment (balbirthomas) [inline tests/test_charm.py:86]: ditto.

**Review by balbirthomas — APPROVED** [review]

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (rbarry82) [reply lib/charms/alertmanager_k8s/v0/consumer.py:41]: It's odd to just have a typing hint on one of the arguments also.

I'd suggest importing `CharmBase` so this looks cleaner and/or also including types for the rest of the non-`self` arguments.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (sed-i) [reply lib/charms/alertmanager_k8s/v0/consumer.py:56]: This reflects my current understanding, and was tested successfully on 2.9.3 and 2.9.5.
I opened an [issue](https://github.com/canonical/alertmanager-operator/issues/8) to refactor this into relation data.

File: src/charm.py
Comment (sed-i) [reply src/charm.py:98]: #49

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (sed-i) [reply lib/charms/alertmanager_k8s/v0/consumer.py:41]: Changed to `CharmBase` and added remaining annotations.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (sed-i) [reply lib/charms/alertmanager_k8s/v0/consumer.py:52]: In that case there will be a "WARNING" entry in the log about incompatibility.
Unless you meant something else?

File: src/charm.py
Comment (sed-i) [reply src/charm.py:40]: Removed

File: src/charm.py
Comment (sed-i) [reply src/charm.py:56]: IIRC, `None` was producing a WARNING in the log. Regardless, it's good to have it versioned.
After we merge the alertmanager PR, I can change this to correspond.

File: src/charm.py
Comment (sed-i) [reply src/charm.py:57]: The `_lib` follows the graylog pattern. I am trying to think of it as a library for prom.

File: src/charm.py
Comment (sed-i) [reply src/charm.py:146]: I intentionally kept changes to the prom op to a minimum, to focus on the interface only.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [reply lib/charms/alertmanager_k8s/v0/consumer.py:52]: ok. so you are leaving that to the base class.That is fine as long as you are aware.

File: src/charm.py
Comment (balbirthomas) [reply src/charm.py:52]: Thanks. I will look at that warning and think about it. Do paste that warning if you will. And if you care try empty string too.

File: src/charm.py
Comment (balbirthomas) [reply src/charm.py:53]: Your call. As I said this is only a style question. I was trying to get you to reflect upon if objects in your programming model compose other objects or libraries ?

File: src/charm.py
Comment (balbirthomas) [reply src/charm.py:142]: Can I request you to create a ticket for future refactor either in prometheus repo or alertmanager repo or both, if you agree that a refactor here may be a good idea. I think it does not make sense to store it in two places. Even more it does not make sense to store it at all. You already have a ticket to that effect (i.e. to use app data).

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [reply lib/charms/alertmanager_k8s/v0/consumer.py:41]: Hmmm... If you must annotate IMHO it is best sparingly done only when required, if type is not obvious, or complex (cf. [non-goals](https://www.python.org/dev/peps/pep-0526/#non-goals))

**Comment by jameinel** [comment]
I think you are misreading non-goals. "Third party packages will have to be
developed to implement such functionality". The non goal is that core
python won't do it, but mypy and other tools *do*.
As do code editors that then let you do autocomplete on attributes, etc.

I believe OF went with all public methods should have type annotations.

John
=:->

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [reply lib/charms/alertmanager_k8s/v0/consumer.py:68]: Of course Prometheus will have to "pull" from consumer lib using `get_cluster_info()`. I mean `ConsumerBase` emits `ProviderAvailable` if the provider and consumer are compatible. How does this check for compatibility factor into the workflow being proposed in this PR. Can it happen that `ConsumerBase` emits `ProviderInvalid` but `AlertmanagerConsumer` emits `ProviderAvailable` in response to the same relation changed event ?

**Comment by balbirthomas** [comment]
> I think you are misreading non-goals. "Third party packages will have to be developed to implement such functionality".  I believe OF went with all public methods should have type annotations. 

I would distinguish between OF, charm and charm library. While a charm has a very sparing "public interface" a charm library does indeed have some. Providing annotations in public interfaces (particularly for complex types) would be reasonable. But mandating it for every string, int etc. I fear will unnecessarily clutter code.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (sed-i) [reply lib/charms/alertmanager_k8s/v0/consumer.py:68]: That is a very good point.
That seems to suggest we should consider a custom event after all?

Comment (balbirthomas) [reply lib/charms/alertmanager_k8s/v0/consumer.py:68]: Custom event would certainly be one way. I am always looking for the "simplest" way. I have not dug deep into this as I trust you are in a better position to do that and trust your judgement. But if you would like to keep things simple kindly do consider if it is possible that the `AlertmanagerConsumer` charm library exposes just one function `get_cluster_info()` and nothing else (no event handler). Since Prometheus just needs the list of alertmangers any time one of the `ConsumerBase` events fire.

File: tests/test_charm.py
Comment (sed-i) [reply tests/test_charm.py]: removed alertmanager test

File: tests/test_charm.py
Comment (sed-i) [reply tests/test_charm.py]: removed alertmanager test

File: src/charm.py
Comment (sed-i) [reply src/charm.py]: removed `_stored.alertmanagers`

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (sed-i) [reply lib/charms/alertmanager_k8s/v0/consumer.py]: I tested without manually emitting, and occasionally there are misses. I'll keep it in for now until the `ready` flag on scale-up/down is resolved.

File: lib/charms/alertmanager_k8s/v0/alertmanager.py
Comment (balbirthomas) [inline lib/charms/alertmanager_k8s/v0/alertmanager.py:116]: Isn't like 116 a duplicate of line 120 or am I missing something ?

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (balbirthomas) [reply lib/charms/alertmanager_k8s/v0/consumer.py]: I do not see how `ready` flag will solve the issue being raised here.

File: lib/charms/alertmanager_k8s/v0/alertmanager.py
Comment (sed-i) [reply lib/charms/alertmanager_k8s/v0/alertmanager.py:116]: Removed.

File: lib/charms/alertmanager_k8s/v0/consumer.py
Comment (dstathis) [reply lib/charms/alertmanager_k8s/v0/consumer.py]: Resolving conversation as a custom event is now used.

File: src/charm.py
Comment (dstathis) [reply src/charm.py]: changed

**Comment by sed-i** [comment]
The changes in this PR are already covered by #65. Closing.

**Comment by sed-i** [comment]
taking over https://github.com/canonical/prometheus-operator/pull/65

File: lib/charms/alertmanager_k8s/v0/alertmanager.py
Comment (balbirthomas) [inline lib/charms/alertmanager_k8s/v0/alertmanager.py:18]: You will need to make another pull request when you finally set the correct LIBID. Since you are in the process of making a bundle. It may be a good idea to finalize your alertmanager library and then merge this PR.

**Review by balbirthomas — APPROVED** [review]

**Pull request merged**
